connaisseur/keys.py

```diff
@@ -0,0 +1,141 @@
+import base64
+import hashlib
+import re
+import binascii
+
+import ecdsa
+import rsa
+from pyasn1.error import PyAsn1Error
+
+from connaisseur.image import Image
+from connaisseur.exceptions import NoSuchClassError, WrongKeyError
+
+KMS_REGEX = r"(?:awskms|gcpkms|azurekms|hashivault|k8s):\/{2,3}[a-zA-Z0-9_.+\/:-]+"
+KEYLESS_REGEX = r"^[a-zA-Z0-9_.+-]+@[a-zA-Z0-9-]+\.[a-zA-Z0-9-.]+$"
+
+
+class Key:
+    """
+    Abstract Key class, containing the public key used for verifying the image signatures.
+    """
+
+    value: str
+
+    def __new__(cls, data: str):
+        try:
+            key_cls = Key.__get_key_type_cls(data)
+            return super(Key, cls).__new__(key_cls)
+        except Exception as err:
+            msg = "Unable to find find class {class_name}."
+            raise NoSuchClassError(message=msg, class_name=key_cls) from err
+
+    @staticmethod
+    def __get_key_type_cls(data: str):
+        # key gets automatically identified
+        if re.match(KEYLESS_REGEX, data):
+            return KeyLessKey
+        elif re.match(KMS_REGEX, data):
+            return KMSKey
+        elif Key.__is_ecdsa_key(data):
+            return ECDSAKey
+        elif Key.__is_rsa_key(data):
+            return RSAKey
+        return None
+
+    @staticmethod
+    def __is_ecdsa_key(data: str):
+        try:
+            ecdsa.VerifyingKey.from_pem(data)
+            return True
+        except (ecdsa.der.UnexpectedDER, binascii.Error, TypeError, AttributeError):
+            return False
+
+    @staticmethod
+    def __is_rsa_key(data: str):
+        try:
+            rsa.PublicKey.load_pkcs1_openssl_pem(data)
+            return True
+        except (ValueError, PyAsn1Error):
+            return False
+
+    def verify(self, validator_type: str, **kwargs):  # pylint: disable=unused-argument
```

Comment:
Something else that occurred to me. Should the key be the object that verifies?

The key itself doesnt really have any connection to what validator is used and hence to how verification is to be done. On the other hand, the validator does know what key to use and how to use said key. While the key doing it doesn't feel bad (I noticed it really late through the review), thinking about it does not feel good either and validator seems like a way better place. Thoughts?

Reply:
I think logically speaking your right. In the current implementation I didn't need to do a differentiation between the keys within the validators, since the keys do the verification, but I see how this is both confusing and logically wrong.

I'll change up the logic so that a validator looks for the key type it's getting and then switches up verification accordingly. Fine with this flow?